wordpress-mobile/WordPress-Android
# Prepublishing Nudges : add visibility option

Fixes #11605

## Findings
Visibility changes allow the user to make their post private, public or password protected. This behavior is connected to the `PostStatus` and the post's `password` field. 

## Solution
The business logic around this implementation goes as follows : 

### Getting Visibility 

1. First I checked to see if the password isn't empty. If it isn't then the `Visibility` is set to `PASSWORD_PROTECTED`. 
2. I then checked to see if the `PostStatus` is equal to `PRIVATE` and if so the `Visibility` is `PRIVATE`. 
3. Based on the design, my assumption was that any other `PostStatus` that is currently set would default to the `PUBLIC` visibility since the post will be accessible with them. 

### Saving Visibility 
1. To transform from a `Visibility` to a `PostStatus` I checked to see if the `Visibility` is set to `PRIVATE` and if so the `PostStatus` would be the same. 
2. For the `PUBLIC` `Visibility` I simply set the `PostStatus` to `DRAFT`. 

### Updating Password

1. To update the password, I utilized the `editPostRepository` and passed the value received from the password dialog used to get the user's input. 
2. In the implementation & tests, I am using an empty string to represent no password, and this is also used to clear it.

### Concerns

1. I noticed that changing the `Visibility`/`PostStatus` will lead to the `PrimaryAction` in the editor being updated and when this happens the action there doesn't trigger the bottom sheet. 

2. The password dialog doesn't have any validation on the `EditText`That is outside the scope of this PR, but something that could be fixed. 

## Testing

1. Create a new post. 
2. Click on `Visibility` in the bottom sheet. 
3. Toggle the private & public visibility options in the list. 
4. Ensure that this is reflected in the `Post Settings` Status.

Public | Private 
--------|-------
<kbd><img src="https://user-images.githubusercontent.com/1509205/80747886-b6f8e780-8ae9-11ea-809c-19a4d863a27a.png" width="320"></kbd>   |  <kbd><img src="https://user-images.githubusercontent.com/1509205/80747993-e1e33b80-8ae9-11ea-9f41-c6e74755ff4d.png" width="320"></kbd>
------------- 



1. Clicking on the password option should trigger a password dialog. 
2. Once you set a password then the `Password Protected` option should be selected.  
3. Go to the `Post Settings` and ensure that the password value was persisted. 
4. You can tap on the `Password Protected` option and clear the password and the visibility will return to `Public`
5. You can verify that the password was successfully cleared in the Post Settings. 


Before | After 
--------|-------
 <kbd><img src="https://user-images.githubusercontent.com/1509205/80748444-8d8c8b80-8aea-11ea-9469-f713f11cd05d.png" width="320"></kbd>  | <kbd><img src="https://user-images.githubusercontent.com/1509205/80748446-8ebdb880-8aea-11ea-8df4-818ebc1330ea.png" width="320"></kbd>
 ------------- 

## Reviewing 
Only 1 reviewer is needed but anyone can review.

## Submitter Checklist
